Performance improvement when uploading big files (100gb++)

gmeks opened the pull request:
When uploading lagre files AppendDataAsync cause alot of Disk IO and CPU usage ( As single chunk can generate over 1k IO).

Test was done with chunk sizes of 10mb.
Added MemorySteam buffer so we do one disk write when the chunk is complete.

SasaCetkovic commented on Source/tusdotnet/Stores/TusDiskStore.cs:
```diff
@@ -32,7 +32,7 @@ public class TusDiskStore :
         // The lower the value, the less data needs to be re-submitted on errors.
         // However, the lower the value, the slower the operation is. 51200 = 50 KB.
-        private const int ByteChunkSize = 5120000;
+        private readonly int ByteChunkSize;
```
should be renamed to _byteChunkSize

gmeks replied:
Done

smatsson commented:
Thanks for the PR! Did you test the code before submitting it? As far as I can tell you never save the file data to disk resulting in the file being lost. 

It seems that you don't quite understand how chunking in the protocol works. Chunking is basically a last resort when the client does not support streaming directly from disk and thus needs smaller chunks to load into memory. Most clients would just send the file as a single "chunk" so reading each chunk into memory before writing to disk would open you up for a denial of service attack, since it's entirely
up to the client to set the chunk size. 

A larger buffer would also require more memory and that the client re-sends more that if the connection is lost. This is obviously a trade off to disk I/O.

I'm not opposed allowing the buffer size to be configurable but it most be done understanding the consequences. I would happily merge a PR for configuring the buffer size given that:

* The code actually work
* There are tests that prove that the code work (e.g. in Source/tusdotnet.test/Tests/TusDiskStoreTests.cs)
* Existing constructors cannot change. Add a new constructor that takes all three parameters. We do not want breaking changes for a feature like this.
* XML comments on the constructor that describe the consequences of changing the buffer size, i.e. larger memory footprint and more data to re-send on errors.

gmeks commented:
Thanks for the check, the code i made on github was a cleaned version of my own version that im testing on. And i forgot a line when moving code over.

I upgraded our (Where i work) custom webside for uploading of large files and in the proccess i switched over from a custom code to TUS based solution ( JS client in browser, and your project server side).

I noticed that uploading large files (100gb++) created a insane CPU/IO usage, so i decided to make a custom version of the library for testing.

Basicly this loop will trigger over 1k times pr chunk.
```
 do
                {
                    if (cancellationToken.IsCancellationRequested)
                    {
                        break;
                    }

                    bytesRead = await stream.ReadAsync(buffer, 0, _byteChunkSize, cancellationToken);

                    fileLength += bytesRead;

                    if (fileLength > uploadLength)
                    {
                        throw new TusStoreException(
                            $"Stream contains more data than the file's upload length. Stream data: {fileLength}, upload length: {uploadLength}.");
                    }

                    chunkBuffer.Write(buffer, 0, bytesRead);
                    bytesWritten += bytesRead;

                } while (bytesRead != 0);
```

So the solution for this is the memory buffer, so we dont have to do any expensive IO operations before the chunk is complete.
There did not seem to be  way to set the chunk size the same as "client" so i added a option for that.

I changed so that the code i run in production is the same as the one in the PR. So all i can say it works for me :)

My code will only write complete chunks to disk, i made the assumtion that this is ok. 
The code is tested by doing a upload, and then verify Sha256 of orginal file and uploaded file.

gmeks commented:
> Existing constructors cannot change. Add a new constructor that takes all three parameters. We do not want breaking changes for a feature like this.

You considered it bad to use default values? Is there a reason?
The overhead in memory would likely been lower if we could use BufferedStream, but not supported by older .net core 

Large files are always split into chunks, i have never seen it not split the files up.

smatsson commented:
I think you are misunderstanding some core concepts of the tus protocol. 
The client will send data to the server and the server will save as much of this data as possible. If the connection dies, the upload can be resumed from the last byte that was saved by the server. I.e. if an upload fails at 99% the client only needs to upload the last 1%. 

Chunking in tus is a way for the client to circumvent some client side/network issues (e.g. memory limitations or proxy timeouts).  Each chunk is still uploaded using the core protocol (i.e send data -> save as much as possible -> resume if connection dies) which makes each chunk resumable as well. 

Chunking is an optional feature on the client and the server can make no assumptions if it's being used or not. The entire file can be transmitted in a single chunk for clients that support that (e.g. the official JS client). 

This PR completely breaks resumability for single chunk uploads and cripples resumability for multi chunk uploads. If the client decides not to use chunks the entire file to be loaded into memory before being saved to disk. This means that the resumability is gone and most likely that the server will end up with a OutOfMemoryException.

The problem with TusDiskStore.AppendDataAsync is that it used the same buffer for reading and writing (e.g. read data -> write data -> read data -> ...). A better solution would be to split the `maxBufferSize` into two; `maxReadBufferSize` and `maxWriteBufferSize`. Each iteration would read (at most) `maxReadBufferSize` from the client into a buffer that is copied to the writeBuffer of size `maxWriteBufferSize`. Once the writeBuffer is full, it will be written to disk and cleared. If the client disconnects, the current data in the writeBuffer will be written to disk.

This would solve your use case by setting a low value for `maxReadBufferSize` (e.g. 50 KB or however much data you feel OK to be re-transmitted if the connection dies) and your client's chunk size as the `maxWriteBufferSize`, while not breaking the code for everyone else. Default buffer sizes should still be 50 KB/50 KB which would give the same functionality as today. 

The only real downside to this compared to the current solution is that the server/API crashes the data is gone. I think this is an acceptable risk.

gmeks commented:
While the protocol may call things streaming, TCP and .NET core 2 does not agree.  The write to disk triggers every 1500 ish bytes ( Basicly around the windows TCP windowing size). This can easly be verified by adding a simple int WriteTriggerCount++ inside the while loop.

I agree with the naming changes, that makes more sense.

I dont see how this would break anything, the only thing im doing is buffering IO.  The previus code could also fail on writing to disk, resoulting in the client having to resend data. Sure the size of what has to be resend might be a bit bigger ( As defined by _maxWriteBufferSize )

The tests included in the project fail at the same rate.

smatsson commented:
Let's try with an example on how this PR breaks stuff. Pretend that the client is sending a 100 GB file and that the JS client is used using the following config:

```
var upload = new tus.Upload(file,
                {
                    endpoint: 'files/',
                    onError: onTusError,
                    onProgress: onTusProgress,
                    onSuccess: onTusSuccess,
                    metadata: {
                        name: file.name,
                        contentType: file.type || 'application/octet-stream'
                    }
                });
```

Current code:
```
Client: Hi, I have this file can I upload data please?
Server: Sure, send it!
Client: <sends data>
Server: Read data from client, write data to disk
Server: Read data from client, write data to disk
Server: Read data from client, write data to disk
(continue until 90 GB is written to disk)
Client: <disconnects>
Client: Hi, I'm back. Could you please provide me the offset for my file so that I can resume it?
Server: Sure, the offset is 90 GB
Client: Cool, here's more data!
Server: Read data from client, write data to disk
...
```

This PR:
```
Client: Hi, I have this file can I upload data please?
Server: Sure, send it!
Client: <sends data>
Server: Reads data from client, write to memory stream. Is chunk complete? No? Then don't write to disk.
Server: Reads data from client, write to memory stream. Is chunk complete? No? Then don't write to disk.
Server: Reads data from client, write to memory stream. Is chunk complete? No? Then don't write to disk.
(continue until 90 GB is written to memory, ignoring any OutOfMemoryException)
Client: <disconnects>
Client: Hi, I'm back. Could you please provide me the offset for my file so that I can resume it?
Server: Sure, the offset is 0.
Client: Ehm what?
...

gmeks commented:
Yea you where correct, if the client actualy streamed a file to the server. It would be buffered entirly in memory. 
I dont know the what senarios make the client/ server decide between streaming vs chunking, i have never seen it actualy stream.

That is fixed now.

smatsson commented:
I tested your code and now it works perfectly. The issues I had with the earlier implementations are now all fixed. Well done! 👍  

There are some minor changes I would like to see done before merging:

I think we should combine `readBufferSize` and int `writeBufferSize` into a single parameter to the `TusDiskStore`. I find this a bit hard to read: 
```
Store = new TusDiskStore(@"C:\tusfiles\", false, 51200, 512000),
```

and would prefer something like:
```
Store = new TusDiskStore(@"C:\tusfiles\", false, new TusDiskStoreBufferSize(51200, 512000)),
```

The comment for the `readBufferSize` parameter reads "This buffer is the maximum amount of data read from the clients stream at once, value should higher then writeBufferSize.". I would argue that this is not correct and that you want this value to be as small as possible and not bigger than `writeBufferSize`. The higher the value, the more data needs to be transmitted if the client disconnects (due to how ASP.NET network streams work).

On line 86 in `TusDiskStore` you create the `writeBuffer` using the `_maxReadBufferSize`. This is probably just a mistake but still want to confirm that this was not intentional? 

On line 370 you re-set the Capacity property of the writeBuffer. This seems unnecessary as the buffer keeps its original capacity? 

We also need unit test coverage. Other tests for the disk store are located in `test\tusdotnet.test\Tests\TusDiskStoreTests.cs`. Please let me know if you need any help with this part.

Remember to add yourself to `~/contributors.md` :)

gmeks commented:
On line 370,
This is based what was done in the most upvoted reply in stackoverflow. So i copied that, assuming that if the capacity is already the same it would have minimal cost.

On line 86 was  a mistake.  FIxed.

The reason for readbuffer and writebuffer being seperate is because it can save a fair bit of memory. For me im currently running it with the following  values:
_maxReadBufferSize = 2000;
_maxWriteBufferSize = 10mb;

Because the while part of AppendDataAsync never realy gets more then 1500-1600 bytes ( My guess is that in .net Core it follows TCP Window in network streams).  So it would be  a waste to generate a 10mb byte array here. But that might behave different in other setups.....  :(

Im going to see if i can make a unite test that captures the generated IO, that could be run on seperate .net versions so perhaps we can remove ReadBufferSize and have a good default.

smatsson commented:
It's a fair point with the extra memory usage for creating a class just to add the two properties. 👍 Just leave it as is. 

I think that the issue you're having with only getting 1500 bytes has something with you setup. I have no issues getting the 50 KB on my machine nor on the servers I've been running it on. Might be IIS or some other setup that needs to be fine tuned. Also noticed that OWIN performs way worse than ASP.NET Core in this matter.

For the unit test you can simply check that AppendDataAsync reads the stream x bytes at a time and that the file is written y bytes at a time. No need to complicate things. There are other tests that does that. 😉

gmeks commented:
You want a spesifc test where i create a new instance of the store?

It seems to that at least the VerifyChecksumAsync test would find data corruption errors. Your more looking for a test that checks it obeys read and write buffer size?

smatsson commented:
Exactly 👍

On Wed, Apr 3, 2019, 13:11 Erling K. Sæterdal <notifications@github.com>
wrote:

> You want a spesifc test where i create a new instance of the store?
>
> It seems to that at least the VerifyChecksumAsync test would find data
> corruption errors. Your more looking for a test that checks it obeys read
> and write buffer size?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tusdotnet/tusdotnet/pull/57#issuecomment-479444590>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYyKSbbxmLSjpGm0O7jqf7fFpjLXR98ks5vdIxEgaJpZM4b_xFy>
> .
>

gmeks commented:
Added one test

smatsson commented:
Nice! Good job. I'll add you to contributors.md and aim to release this with 2.2
